urls: Generate narrow links using "channel" operator in web app and server.

laurynmm opened the pull request:
Updates the web app and server to generate links using the "channel" operator instead of the deprecated "stream" operator.

Fixes #30385.

**Notes**:
- Prep commit to update any existing links in our documentation that use the "stream" operator.
- Existing links with the "stream" operator continue to work as expected in the web app, and the URL in the browser is not updated when one navigates to those links.
- `git grep -c '#narrow/stream'` returns two results:
  - `zerver/migrations/0501_delete_dangling_usermessages.py:1`: Print logging strings of near message URLs in this migration.
  - `zerver/tests/fixtures/rocketchat_fixtures/rocketchat_message.bson:3`: Some chat.zulip.org links that were included in messages in the export data fixtures to test the import code.
- In the commit with frontend changes, updates a few comments in `web/src/filter.ts` that have been added since the channel to stream rename happened. It's useful for that file in particular to use "channel" unless referring to the message type or an imported module/function.
- ~Final commit updates message fetching to use the channel ID instead of the channel name in the narrow, if it exists.~

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

laurynmm commented on web/src/message_fetch.ts:
```diff
@@ -235,6 +236,7 @@ function handle_operators_supporting_id_based_api(data: {
     const operators_supporting_id = new Set([
         "id",
         "stream",
+        "channel",
```
I'm pretty sure that the web app is only going to have "channel" here. But if we want to keep this aligned with the same `operators_supporting_id` in `zerver/lib/narrow.py`, then these changes make sense. If we're fine with those two sets being different, then we can just have "channel" here and check for that one string below.

timabbott replied:
I think it's fine for these to be different from the backend if we expect it'll have been canonicalized to `channel` before getting here. We could consider changing the server-side implementation to canonicalize early too.

laurynmm replied:
I put up what I think is a better fix for this in #31934, which refactors `handle_operators_support_id_based_api` and `loading_messages`. Since this is somewhat tangential to the changes to generate URLs with the "channel" operator, I'm going to drop this commit from this PR.

laurynmm commented:
**Updates**:
- Dropped final commit that updated `message_fetch.handle_operators_supporting_id_based_api` for the "channel" operator. That's now addressed separately in #31934.

timabbott commented on zerver/lib/url_decoding.py:
```diff
@@ -23,4 +23,4 @@ def is_same_server_message_link(url: str) -> bool:
     ends_with_near_message_id = fragment_parts[-2] == "near" and fragment_parts[-1].isdigit()
 
-    return category == "narrow" and section in {"stream", "dm"} and ends_with_near_message_id
+    return category == "narrow" and section in {"channel", "dm"} and ends_with_near_message_id
```
Hmm, this function might be used in a way that we should be supporting the legacy values too, in that drafts may still have the old syntax. Maybe I can just add a bonus commit for that detail.

timabbott replied:
Thinking about this a bit more, I think it's likely to be sufficiently rare that this actually happens that it's not worth introducing temporary legacy code around it, given the impact will be just a bit less nice desktop/push notification alert.

timabbott commented:
Merged, thanks @laurynmm! splitting the more controversial change to another PR was definitely helpful.
